### PR TITLE
Move Lazy Shape Inference functions to pytorch core

### PR DESCRIPTION
**Pull request by wconstab**
Test Plan: tested via lazy_tensor_staging branch (test_ptltc operator tests as well as torchbench inference)

Differential Revision: D34187999

### CODE REVIEW

File: tools/codegen/api/lazy.py
Comment (bdhirsh) [inline tools/codegen/api/lazy.py:66]: Is this meant to fix https://github.com/pytorch/pytorch/issues/72852? (It looks like `isValueType()` never actually gets called with a normal `Type` if I'm following correctly, which makes this an easy fix)

File: tools/codegen/api/lazy.py
Comment (bdhirsh) [inline tools/codegen/api/lazy.py:74]: I don't totally follow from the comment - can you give a code example?

It also doesn't seem like this really matches the description of the function: "This is equivalent to being Tensor-like", if we directly return true on scalar types.

File: tools/codegen/api/lazy.py
Comment (wconstab) [reply tools/codegen/api/lazy.py:66]: yes, it is (to both)

File: tools/codegen/api/lazy.py
Comment (bdhirsh) [inline tools/codegen/api/lazy.py:139]: Basic question - it looks like right now, `LazyIrSchema` has separate fields for positional and keyword arguments. Is that info used meaningfully anywhere?

If not, I think you could simplify a bunch of this logic into something like:
```
self.keyword_arg_types = [NamedCType(arg.name, process_ir_type(arg.type)) for ag in func.schema_order_arguments()]
```

That would help you avoid having to special-case `TensorOptionsArguments`, `SelfArgument`, and have to explicitly list out the `Arguments` field names (which I think would cause LTC codegen to break in pretty subtle ways if we changed the names of those fields in core codegen)

File: tools/codegen/api/lazy.py
Comment (bdhirsh) [inline tools/codegen/api/lazy.py:143]: Ok, so the `LazyIrSchema` object knows "what are the C++ types and names of all arguments in my schema", and now it also knows "which of my arguments were originally scalars before processing", using this separate list.

I think the other "data-model-y" option you could take would be to represent that info per-argument using a new type (e.g. a `LTCArgument` that knows both its `NamedCType`, and whether it was a scalar argument). The advantage there is that your processing functions (`node_ctor_arg_rvalue_string`) can run on each argument in isolation, instead of having to take in the whole schema. But if this list is used in a really minimal/targeted way then maybe that's overkill.

File: tools/codegen/dest/lazy_ir.py
Comment (bdhirsh) [inline tools/codegen/dest/lazy_ir.py:71]: side note: now that this is all in core, should we just add these symbols in core instead of having to handle them specially in codegen?

File: tools/codegen/dest/lazy_ir.py
Comment (bdhirsh) [inline tools/codegen/dest/lazy_ir.py:307]: nit: commented code

File: tools/codegen/dest/lazy_ir.py
Comment (bdhirsh) [inline tools/codegen/dest/lazy_ir.py:320]: This doesn't seem like a big deal - but any reason why we need to use the `metadata.kernel` to choose how to name the shape functions? For example, we could do something like `f.func.name.unambiguous_name()` instead

File: tools/codegen/dest/lazy_ir.py
Comment (bdhirsh) [inline tools/codegen/dest/lazy_ir.py:232]: what's the TODO for?

File: tools/codegen/dest/lazy_ir.py
Comment (bdhirsh) [inline tools/codegen/dest/lazy_ir.py:189]: did we lose the assert?

Comment (bdhirsh) [reply tools/codegen/dest/lazy_ir.py:189]: (NVM, i see the new one further up)

File: tools/codegen/dest/lazy_ir.py
Comment (bdhirsh) [inline tools/codegen/dest/lazy_ir.py:265]: Just confirming - so this PR also adds "per-op-defined eager fallback"?

File: tools/codegen/gen_lazy_tensor.py
Comment (bdhirsh) [inline tools/codegen/gen_lazy_tensor.py:178]: Do you mind writing out somewhere what the flow looks like for non-meta tensor shape inference? It'd help to have it listed somewhere:
(1) what all of the components are (the headers, shape inference kernels, their call-sites)
(2) which parts are codegen'd

File: torch/csrc/lazy/core/shape_inference.cpp
Comment (bdhirsh) [inline torch/csrc/lazy/core/shape_inference.cpp:137]: oh, I think you might be hitting CI failures because `toValueType()` was recently changed to `toRealValueType()`

File: torch/csrc/lazy/core/shape_inference.cpp
Comment (wconstab) [reply torch/csrc/lazy/core/shape_inference.cpp:14]: I could put a ref to read the 'tech debt' section in shape_inference.cpp in the codegen file.

I'm not sure what you mean about inplace codegen though?

File: torch/csrc/lazy/core/shape_inference.cpp
Comment (wconstab) [reply torch/csrc/lazy/core/shape_inference.cpp:137]: ahh, gotcha.  I will follow up on that. hopefully it is just a trivial rename otherwise i'll have to do some gymnastics to get toRealValueType on staging branch first and test it there..

File: tools/codegen/dest/lazy_ir.py
Comment (wconstab) [reply tools/codegen/dest/lazy_ir.py:71]: I actually added a handful of symbols to core, and was never sure why i had to do that, and ultimately figured this was a cleaner escape hatch.

if the idea is to have all the symbols be constants in core, should we just make a diff that adds all the symbols at once?  otherwise, it's not super clear why we have some and not others...

File: tools/codegen/dest/lazy_ir.py
Comment (wconstab) [reply tools/codegen/dest/lazy_ir.py:232]: ah that should be removed..

File: tools/codegen/dest/lazy_ir.py
Comment (wconstab) [reply tools/codegen/dest/lazy_ir.py:265]: the fallback thing here is just a debugging feature.  It lets you set an env variable that will force a particular operator to fall back.

File: tools/codegen/dest/lazy_ir.py
Comment (wconstab) [reply tools/codegen/dest/lazy_ir.py:320]: I can't answer off the top of my head.  I probably didn't know about unambiguous_name().  I think i needed metadata.kernel to overcome some ambiguity so this may be possible to substitute. I'll have to test it separately (back on staging) and then drop the update in if it works.

Comment (wconstab) [reply tools/codegen/dest/lazy_ir.py:320]: ok, so here is one case where they differ.

ipdb> p metadata.kernel
'arange_out'
ipdb> p f.func.name.unambiguous_name()
'arange_start_out'

I'm not sure what the right thing to do is.

File: tools/codegen/api/lazy.py
Comment (wconstab) [reply tools/codegen/api/lazy.py:74]: lazy tensor tracing promotes at::Scalars to become lazy tensors.  This enables treating the scalars as inputs to the lazy trace and not burning them into the program.

it could use a bit of fresh paint.  i think i might want to refactor some of this stuff but probably in a separate PR.  I could use your suggestions for that too more generally.   but I don't think I want to make too many changes at this PR since it's aiming to just sync up master to what we are already using on staging.  Better to develop the new version on staging where we can test it directly.  (until we finally finish merging enough to master that we can build/run entirely on master's CI.)

File: tools/codegen/api/lazy.py
Comment (wconstab) [reply tools/codegen/api/lazy.py:139]: if you look at `lazy_ts_lowering.py` you'll see how we generate different code for kw args vs positional args.  I couldn't think of a cleaner way at the time but I agree it's a bit complex.

File: tools/codegen/api/lazy.py
Comment (wconstab) [reply tools/codegen/api/lazy.py:143]: i think this approach sounds good.  generally i think i want to separate out more 'refactoring' to a separate diff.  Lets collect ideas for that on a GH issue. i'll make one and link here: https://github.com/pytorch/pytorch/issues/72960

File: tools/codegen/dest/lazy_ir.py
Comment (bdhirsh) [reply tools/codegen/dest/lazy_ir.py:71]: I was trying to look into where `at::aten::{symbol}` is actually defined - do you know where it is?

 eter Bell landed some changes recently that I think are meant to codegen all of the aten symbols, so I'm wondering if this probably was basically already solved for you. (I can see `sigmoid_backward` listed in the codegen'd file `build/aten/src/ATen/core/aten_interned_strings.h`)

File: tools/codegen/dest/lazy_ir.py
Comment (bdhirsh) [reply tools/codegen/dest/lazy_ir.py:319]: leaving for a refactor sounds fine.

I think at a high level, what we probably want to have happen (based on what some of the other codegen in core does), is that the shape signatures are part of some API (the "lazy API", or "lazy shape API") that knows how to generate its name. That way at both the place where you define those functions in the codegen, and at the place where you call those functions, you can use the same api call to generate the name (and you don't have to remember whether or not it was actually using `metadata.kernel` or `f.unambiguous_name()`.

**Review by bdhirsh — APPROVED** [review]
Approving, since the main purpose of this PR is to true up master with the staging branch and refactors are left as a followup item in https://github.com/pytorch/pytorch/issues/72960

File: tools/codegen/dest/lazy_ir.py
Comment (wconstab) [reply tools/codegen/dest/lazy_ir.py:319]: yea that makes sense. i'll reflect that in the other (refactor) issue.

**Pull request closed without merging**
